### PR TITLE
feat(vtxo-manager): per-call threshold on renewVtxos

### LINKED ISSUE #388: Allow setting custom threshold when running `VtxoManager.renewVtxos()`
## Context

`VtxoManager.renewVtxos()` does not accept a custom threshold argument, making it impossible to renew only VTXOs that are urgently expiring (beyond the globally configured threshold) without reconstructing the manager with a different config.

The underlying `getExpiringVtxos(thresholdMs?: number)` already accepts an optional override, but `renewVtxos()` uses either `settlementConfig.vtxoThreshold` or `DEFAULT_RENEWAL_CONFIG.thresholdMs` (72 hours) — with no way for callers to pass a different threshold at call time.

The [VTXO management docs](https://docs.arkadeos.com/wallets/v0.4/advanced/vtxo-management.mdx) previously showed a threshold argument for `renewVtxos()` and `getExpiringVtxos()`, but this no longer matches the actual API (and the docs now reference a stale `thresholdPercentage` parameter that doesn't exist in code — a separate docs cleanup may be warranted).

---

## Affected Components

- **`src/wallet/vtxo-manager.ts`** — `renewVtxos()` at line 625, `getExpiringVtxos()` at line 567
- **Interface** — `IVtxoManager.renewVtxos()` (line 368): signature must be updated in parallel
- **Docs** — `wallets/v0.4/advanced/vtxo-management.mdx` in `arkade-os/docs` (references stale `thresholdPercentage` and an old `renewVtxos()` signature)

---

## Root Cause

`renewVtxos()` hardcodes the threshold resolution:
```typescript
// vtxo-manager.ts ~L637
const vtxos = await this.getExpiringVtxos(
    this.settlementConfig !== false &&
        this.settlementConfig?.vtxoThreshold !== undefined
        ? this.settlementConfig.vtxoThreshold * 1000
        : DEFAULT_RENEWAL_CONFIG.thresholdMs
);
```

`getExpiringVtxos` already supports an override parameter; `renewVtxos` simply doesn't expose it.

---

### Option A — Add `thresholdMs` param (matches existing internal API)
```typescript
async renewVtxos(
    eventCallback?: (event: SettlementEvent) => void,
    thresholdMs?: number
): Promise<string>
```
Pass `thresholdMs` through to `getExpiringVtxos()` when provided.

### Option B — Add `thresholdSeconds` param (matches `SettlementConfig.vtxoThreshold` units)
```typescript
async renewVtxos(
    eventCallback?: (event: SettlementEvent) => void,
    options?: { thresholdSeconds?: number }
): Promise<string>
```
Converts to ms before passing to `getExpiringVtxos()`. More consistent with the rest of the public API.

Option B is preferred for consistency with `SettlementConfig.vtxoThreshold` (which uses seconds), though either approach works. An `options` object future-proofs the API better than positional params.

---

## Changes Required

1. **`src/wallet/vtxo-manager.ts`**
   - Update `renewVtxos()` signature to accept optional threshold
   - Update the `IVtxoManager` interface accordingly
   - Pass the threshold override to `getExpiringVtxos()`

2. **`arkade-os/docs`** — `wallets/v0.4/advanced/vtxo-management.mdx`
   - Update code examples to reflect actual `renewVtxos()` and `getExpiringVtxos()` signatures
   - Replace stale `thresholdPercentage` references with actual `thresholdMs` / `thresholdSeconds` param
   - Document the new threshold override usage

---

## Related

- `getExpiringVtxos(thresholdMs?)` already supports threshold override — this is the underlying fix point
- `SettlementConfig.vtxoThreshold` (seconds) vs `RenewalConfig.thresholdMs` (ms) — the existing unit inconsistency between legacy and current config; new param should follow `SettlementConfig` convention (seconds)
- `DEFAULT_THRESHOLD_SECONDS = 259200` (3 days) — the fallback used when no threshold is configured

### PR DESCRIPTION
**Pull request by pietro909**
## Summary

`IVtxoManager.renewVtxos()` now accepts an optional `{ thresholdSeconds }` second argument that overrides the renewal threshold for that call only. Useful when a caller wants to renew only VTXOs that are more urgently expiring than the globally-configured `SettlementConfig.vtxoThreshold`.

Resolution order:

1. `options.thresholdSeconds` (new)
2. `SettlementConfig.vtxoThreshold`
3. `DEFAULT_THRESHOLD_SECONDS` (3 days)

The new option is propagated through the ServiceWorker message bus.

Closes #388.

## Changes

- `src/wallet/vtxo-manager.ts`: new `RenewVtxosOptions` type; updated `IVtxoManager.renewVtxos` + `VtxoManager.renewVtxos`.
- `src/wallet/serviceWorker/wallet-message-handler.ts`: `RequestRenewVtxos` carries optional payload; handler forwards it.
- `src/wallet/serviceWorker/wallet.ts`: SW `VtxoManager` proxy forwards options in the message.
- `src/index.ts`: re-exports `RenewVtxosOptions`.
- `test/wallet-message-handler.test.ts`: assertion updated for the new arg.

## Test plan

- [x] `pnpm run lint` clean
- [x] `pnpm -C packages/ts-sdk exec tsc --noEmit` clean
- [x] `pnpm run test:unit` — 1228/1228 pass (1 pre-existing skip)

## Summary by CodeRabbit

* **New Features**
  * Per-call renewal threshold override: you can now pass an optional threshold when triggering VTXO renewal to customize behavior without changing global settings.

* **Bug Fixes**
  * Service-worker request/handling updated to forward optional renewal options correctly.
  * Invalid threshold values now raise a clear error instead of proceeding silently.

* **Tests**
  * Coverage added for optional threshold payload and call signatures.

[Review Change Stack](https://app.coderabbit.ai/change-stack/arkade-os/ts-sdk/pull/512?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by pietro909** [comment]
@coderabbitai review please

File: packages/ts-sdk/src/wallet/vtxo-manager.ts [hunk 1]
```diff
@@ -715,12 +739,19 @@ export class VtxoManager implements AsyncDisposable, IVtxoManager {
...
         try {
             // Get all virtual outputs (including recoverable ones)
-            // Use default threshold to bypass settlementConfig gate (manual API should always work)
-            const threshold =
+            // Resolution order: explicit options.thresholdSeconds > settlementConfig.vtxoThreshold > default.
+            // Manual API should always work, so we bypass the settlementConfig === false gate.
+            let threshold: number;
+            if (options?.thresholdSeconds !== undefined) {
+                threshold = options.thresholdSeconds * 1000;
+            } else if (
```
[the remark this answers is not in the document]
Comment (gringokiwi) [reply packages/ts-sdk/src/wallet/vtxo-manager.ts:766]: Could be a good idea @pietro909

**Review by gringokiwi — APPROVED** [review]
Other than the CodeRabbit suggestion lgtm

**Comment by bordalix** [comment]
Fix conflicts + rabbit suggestion.
If tests are green, merge it.

**Review by ghost — CHANGES_REQUESTED** [review]
## Arkana Code Review — `feat(vtxo-manager): per-call threshold on renewVtxos`

Clean, well-scoped change. The plumbing through the ServiceWorker message bus is correct. Two issues to address before merge.

---

### 🔴 Missing input validation on `thresholdSeconds` (request changes)

**`vtxo-manager.ts` (new threshold resolution block, ~line 739–747 in the diff)**

`options.thresholdSeconds` is multiplied by 1000 and passed to `getExpiringVtxos` → `isVtxoExpiringSoon`. There is no validation that it's a positive finite number.

Edge cases:
- **Negative value** → negative `thresholdMs` → hits the `<= 100` guard in `isVtxoExpiringSoon:339` → silently falls back to `DEFAULT_THRESHOLD_MS` (3 days). Caller thinks they're using a custom threshold but gets the global default. Silent wrong behavior.
- **Zero** → same silent fallback to 3 days.
- **`NaN` / `Infinity`** → `NaN * 1000 = NaN`, `NaN <= 100` is false → `batchExpiry - now <= NaN` is always false → no VTXOs match → throws "No VTXOs available to renew". Confusing error for the caller.
- **Very small positive values** (e.g. `0.05` = 50ms threshold) → also silently overridden by the `<= 100` guard.

**Fix:** Validate at the top of `renewVtxos`:
```ts
if (options?.thresholdSeconds !== undefined) {
    if (!Number.isFinite(options.thresholdSeconds) || options.thresholdSeconds <= 0) {
        throw new Error("thresholdSeconds must be a positive finite number");
    }
}
```

The `isVtxoExpiringSoon` guard at line 339 is a pre-existing footgun for any sub-100ms threshold, but that's out of scope for this PR. Validating at the entry point is sufficient.

---

### 🟡 No unit tests for the new code path (blocking)

**`test/vtxo-manager.test.ts`** — The `renewVtxos` describe block has 5 tests, none exercise `thresholdSeconds`. The only test change is in `wallet-message-handler.test.ts` updating the assertion from `(expect.any(Function))` to `(expect.any(Function), undefined)`.

Needed tests:
1. `renewVtxos(undefined, { thresholdSeconds: X })` uses the provided threshold instead of `settlementConfig.vtxoThreshold`
2. `renewVtxos(undefined, { thresholdSeconds: X })` overrides even when `settlementConfig.vtxoThreshold` is set
3. Invalid `thresholdSeconds` (negative, zero, NaN) throws immediately
4. ServiceWorker message handler forwards `payload` to `renewVtxos` when present

---

### ✅ Looks good

- **Backward compatibility**: New `options` param is optional and appended as 2nd arg. All existing callers (`wallet/src/providers/wallet.tsx:617`, all tests, e2e) pass 0 or 1 args — no breakage.
- **Resolution order** (explicit > config > default) is correct and well-documented in the code comment.
- **ServiceWorker plumbing**: `RequestRenewVtxos.payload` → handler extracts and passes to `renewVtxos` — correct.
- **Type export**: `RenewVtxosOptions` correctly exported from `src/index.ts`.
- **Cross-repo impact**: Checked `wallet`, `arkadian`, `demos`, `wdk-demo-app`, and other repos under `/root/arkana/repos/`. No consumers pass a second argument today.

---

⚠️ **Human review requested** — this touches VTXO renewal lifecycle (which VTXOs get renewed and when). Not protocol-critical (no signing/forfeit/exit changes), but the threshold controls what gets settled, so incorrect behavior here could leave VTXOs unrenewed and at risk of expiry.

🤖 Reviewed by Arkana

**Review by ghost — CHANGES_REQUESTED** [review]
## Arkana Code Review — `feat(vtxo-manager): per-call threshold on renewVtxos`

Clean, well-scoped change. The plumbing through the ServiceWorker message bus is correct. Two issues to address before merge.

---

### 🔴 Missing input validation on `thresholdSeconds` (request changes)

**`vtxo-manager.ts` (new threshold resolution block, ~line 739–747 in the diff)**

`options.thresholdSeconds` is multiplied by 1000 and passed to `getExpiringVtxos` → `isVtxoExpiringSoon`. There is no validation that it's a positive finite number.

Edge cases:
- **Negative value** → negative `thresholdMs` → hits the `<= 100` guard in `isVtxoExpiringSoon:339` → silently falls back to `DEFAULT_THRESHOLD_MS` (3 days). Caller thinks they're using a custom threshold but gets the global default. Silent wrong behavior.
- **Zero** → same silent fallback to 3 days.
- **`NaN` / `Infinity`** → `NaN * 1000 = NaN`, `NaN <= 100` is false → `batchExpiry - now <= NaN` is always false → no VTXOs match → throws "No VTXOs available to renew". Confusing error for the caller.
- **Very small positive values** (e.g. `0.05` = 50ms threshold) → also silently overridden by the `<= 100` guard.

**Fix:** Validate at the top of `renewVtxos`:
```ts
if (options?.thresholdSeconds !== undefined) {
    if (!Number.isFinite(options.thresholdSeconds) || options.thresholdSeconds <= 0) {
        throw new Error("thresholdSeconds must be a positive finite number");
    }
}
```

The `isVtxoExpiringSoon` guard at line 339 is a pre-existing footgun for any sub-100ms threshold, but that's out of scope for this PR. Validating at the entry point is sufficient.

---

### 🟡 No unit tests for the new code path (blocking)

**`test/vtxo-manager.test.ts`** — The `renewVtxos` describe block has 5 tests, none exercise `thresholdSeconds`. The only test change is in `wallet-message-handler.test.ts` updating the assertion from `(expect.any(Function))` to `(expect.any(Function), undefined)`.

Needed tests:
1. `renewVtxos(undefined, { thresholdSeconds: X })` uses the provided threshold instead of `settlementConfig.vtxoThreshold`
2. `renewVtxos(undefined, { thresholdSeconds: X })` overrides even when `settlementConfig.vtxoThreshold` is set
3. Invalid `thresholdSeconds` (negative, zero, NaN) throws immediately
4. ServiceWorker message handler forwards `payload` to `renewVtxos` when present

---

### ✅ Looks good

- **Backward compatibility**: New `options` param is optional and appended as 2nd arg. All existing callers (`wallet/src/providers/wallet.tsx:617`, all tests, e2e) pass 0 or 1 args — no breakage.
- **Resolution order** (explicit > config > default) is correct and well-documented in the code comment.
- **ServiceWorker plumbing**: `RequestRenewVtxos.payload` → handler extracts and passes to `renewVtxos` — correct.
- **Type export**: `RenewVtxosOptions` correctly exported from `src/index.ts`.
- **Cross-repo impact**: Checked `wallet`, `arkadian`, `demos`, `wdk-demo-app`, and other repos under `/root/arkana/repos/`. No consumers pass a second argument today.

---

⚠️ **Human review requested** — this touches VTXO renewal lifecycle (which VTXOs get renewed and when). Not protocol-critical (no signing/forfeit/exit changes), but the threshold controls what gets settled, so incorrect behavior here could leave VTXOs unrenewed and at risk of expiry.

🤖 Reviewed by Arkana

File: packages/ts-sdk/src/wallet/vtxo-manager.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (pietro909) [reply packages/ts-sdk/src/wallet/vtxo-manager.ts:766]: fixed in e0a50c6 @coderabbitai please review it again

**Review by ghost — APPROVED** [review]
## Arkana Code Review — follow-up on `e0a50c6b`

Previous review requested two changes. Status:

### ✅ Validation — resolved

The new validation block in `vtxo-manager.ts:733–745` is thorough and correctly placed:
- Runtime type check (`typeof !== "number"`) — good, since payloads cross the MessageBus trust boundary
- `!Number.isFinite()` catches NaN/Infinity
- `<= 0` catches negative/zero
- `TypeError` thrown **before** `renewalInProgress` is set — no state corruption on invalid input
- Comment clearly explains the "why"

This fully addresses the silent-wrong-behavior edge cases I flagged (negative → silent 3-day fallback, NaN → confusing "no VTXOs" error).

### 🟡 Tests — partially resolved (non-blocking)

New test in `wallet-message-handler.test.ts` covers ServiceWorker forwarding with `{ thresholdSeconds }` payload — good.

Still missing from `vtxo-manager.test.ts` (suggest adding in a follow-up):
- Validation rejects invalid inputs (negative, zero, NaN, string) with `TypeError`
- `thresholdSeconds` overrides `settlementConfig.vtxoThreshold` when both are set
- Default fallback when neither is set

These are low-risk given the validation logic is straightforward, so not blocking.

---

⚠️ **Human review still requested** — VTXO renewal lifecycle. Same rationale as before: threshold controls which VTXOs get renewed, incorrect behavior risks expiry.

🤖 Reviewed by Arkana
